### ISSUE TITLE
compaction_additional_test.py::TestLCSSSTablePromotion::test_lcs_table_promotion_after_stcs_migration fails

### ISSUE THREAD

**Issue by tchaikov** [scylladb/scylladb#15405]
```
13:39:35  =================================== FAILURES ===================================
13:39:35  ____ TestLCSSSTablePromotion.test_lcs_table_promotion_after_stcs_migration _____
13:39:35  [gw3] linux -- Python 3.11.4 /usr/bin/python3
13:39:35  
13:39:35  self = <compaction_additional_test.TestLCSSSTablePromotion object at 0x7f597c6bdd50>
13:39:35  
13:39:35      def test_lcs_table_promotion_after_stcs_migration(self):
13:39:35          node, session, storage_service_client = self._prepare()
13:39:35          create_ks(session=session, name=self.KS, rf=1)
13:39:35          create_cf(session=session, name=self.CF, columns={'c1': 'text', 'c2': 'text'},
13:39:35                    compaction=self.STCS)
13:39:35          keys_to_insert = [20_000, 80_000, 150_000, 250_000, 500_000]
13:39:35      
13:39:35          for item in keys_to_insert:
13:39:35              insert_c1c2(session=session, n=item)
13:39:35              node.flush()
13:39:35      
13:39:35          session.execute(f"ALTER TABLE ks.cf WITH compaction={self.LCS}")
13:39:35          node.nodetool(f"refresh {self.KS} {self.CF}")
13:39:35      
13:39:35          regex_match = self._get_table_levels(node=node)
13:39:35  >       self._validate_levels_distribution(regex_match)
...
13:39:35  compaction_additional_test.py:2775: 
13:39:35  _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
13:39:35  
13:39:35  levels_regex = <re.Match object; span=(199, 236), match='SSTables in each level: [0, 20/10, 2]'>
13:39:35  
13:39:35      @staticmethod
13:39:35      def _validate_levels_distribution(levels_regex: Match[AnyStr]):
13:39:35          assert levels_regex
13:39:35      
13:39:35          # Get the sstable_list from the Match object and parse it into a list of integers
13:39:35          raw_levels = [levels_regex.groupdict().get("sstable_list").split(",")][0]
13:39:35      
13:39:35          levels = [int(item.split('/')[0] if '/' in item else item) for item in raw_levels]
13:39:35          assert levels[-1] != sum(levels), "Expected sstables to not be promoted solely to the " \
13:39:35                                            "top level, but found all in the top level: %s" % raw_levels
13:39:35      
13:39:35          level_count_validation = reduce(lambda x, y: y >= (x * 10), levels)
13:39:35  >       assert level_count_validation, "Expected each LCS level to be at least 10x of the previous " \
13:39:35                                         "level, but they were not: %s" % raw_levels
13:39:35  E       AssertionError: Expected each LCS level to be at least 10x of the previous level, but they were not: ['0', ' 20/10', ' 2']
13:39:35  E       assert False
```

see 

- https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/4503/
- https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/4504/
- https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/4505/

**Comment by tchaikov** [comment]
[logs-full.release.005.zip](https://github.com/scylladb/scylladb/files/12605664/logs-full.release.005.zip)

**Comment by tchaikov** [comment]
the failing test is not a new one -- it was added more one year ago: https://github.com/scylladb/scylla-dtest/commit/3013f2baceab60c97cd8e4ef099cd16ec350bab7

**Comment by tchaikov** [comment]
i am trying to reproduce it locally, but i ended up with another failure from this test
```
FAILED compaction_additional_test.py::TestLCSSSTablePromotion::test_lcs_table_promotion_after_stcs_migration - AssertionError: Expected sstables to not be promoted solely to the top level, but found all in the top level: ['6/4']
```

**Comment by mykaul** [comment]
What's the next step here then?

**Comment by fruch** [comment]
happened on enterprise/next as well

https://jenkins.scylladb.com/job/scylla-enterprise/job/next/2576/testReport/compaction_additional_test/TestLCSSSTablePromotion/Tests___Sanity_Tests_RAFT___test_lcs_table_promotion_after_stcs_migration/

next step would be to get @raphaelsc to take a look.

**Comment by avikivity** [comment]
Seen again:

https://jenkins.scylladb.com/view/nexts/job/scylla-master/job/next/6621/testReport/junit/compaction_additional_test/TestLCSSSTablePromotion/Tests___Sanity_Tests___test_lcs_table_promotion_after_stcs_migration/

Error Message
AssertionError: Expected sstables to not be promoted solely to the top level, but found all in the top level: ['2']
assert 2 != 2
 +  where 2 = sum([2])
Stacktrace
self = <compaction_additional_test.TestLCSSSTablePromotion object at 0x7fba766c0850>

    def test_lcs_table_promotion_after_stcs_migration(self):
        node, session, storage_service_client = self._prepare()
        create_ks(session=session, name=self.KS, rf=1)
        create_cf(session=session, name=self.CF, columns={'c1': 'text', 'c2': 'text'},
                  compaction=self.STCS)
        keys_to_insert = [20_000, 80_000, 150_000, 250_000, 500_000]
    
        for item in keys_to_insert:
            insert_c1c2(session=session, n=item)
            node.flush()
    
        session.execute(f"ALTER TABLE ks.cf WITH compaction={self.LCS}")
        node.nodetool(f"refresh {self.KS} {self.CF}")
    
        regex_match = self._get_table_levels(node=node)
>       self._validate_levels_distribution(regex_match)

compaction_additional_test.py:2775: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

levels_regex = <re.Match object; span=(211, 238), match='SSTables in each level: [2]'>

    @staticmethod
    def _validate_levels_distribution(levels_regex: Match[AnyStr]):
        assert levels_regex
    
        # Get the sstable_list from the Match object and parse it into a list of integers
        raw_levels = [levels_regex.groupdict().get("sstable_list").split(",")][0]
    
        levels = [int(item.split('/')[0] if '/' in item else item) for item in raw_levels]
>       assert levels[-1] != sum(levels), "Expected sstables to not be promoted solely to the " \
                                          "top level, but found all in the top level: %s" % raw_levels
E       AssertionError: Expected sstables to not be promoted solely to the top level, but found all in the top level: ['2']
E       assert 2 != 2
E        +  where 2 = sum([2])

compaction_additional_test.py:2814: AssertionError

**Comment by tchaikov** [comment]
i will give it a try tomorrow.

**Comment by avikivity** [comment]
Thanks. @raphaelsc if you have ideas/pointers, please share.

**Comment by avikivity** [comment]
Seen again: https://jenkins.scylladb.com/view/nexts/job/scylla-master/job/next/6627/

**Comment by raphaelsc** [comment]
the dtest is built on the assumption that LCS has level L 10x bigger than L-1 as an invariant (meaning that would be the layout at any point in time), but that's not true today. That layout is eventually reached through promotions, but the dtest is not giving compaction enough time to complete its work.

the log shows compaction was interrupted 

```
INFO  2023-09-14 05:15:06,005 [shard 0:comp] compaction - [Compact ks.cf aa9ebb50-52bd-11ee-8637-ad387c44b32e] Compacting of 1 sstables interrupted due to: sstables::compaction_stopped_exception (Compaction for ks/cf was stopped due to: shutdown)
INFO  2023-09-14 05:15:06,005 [shard 0:comp] compaction_manager - Compaction task 0x600001412710 for table ks.cf compaction_group=0/1 [0x60000147cd40]: Compaction for ks/cf was stopped due to: shutdown: stopping
```

A simple correction is about waiting for compaction to cease, before making the checks.

**Comment by raphaelsc** [comment]
> i am trying to reproduce it locally, but i ended up with another failure from this test
> 
> ```
> FAILED compaction_additional_test.py::TestLCSSSTablePromotion::test_lcs_table_promotion_after_stcs_migration - AssertionError: Expected sstables to not be promoted solely to the top level, but found all in the top level: ['6/4']
> ```
> 
> and
> 
> ```
> FAILED compaction_additional_test.py::TestLCSSSTablePromotion::test_lcs_table_promotion_after_stcs_migration - AssertionError: Expected sstables to not be promoted solely to the top level, but found all in the top level: ['8/4']
> ```

This shows Kefu's machine is too fast, and the checks were performed before LCS had even chance to promote any sstable to higher level. My fix suggestion above applies here too.

**Comment by fruch** [comment]
> the dtest is built on the assumption that LCS has level L 10x bigger than L-1 as an invariant (meaning that would be the layout at any point in time), but that's not true today. That layout is eventually reached through promotions, but the dtest is not giving compaction enough time to complete its work.
> 
> the log shows compaction was interrupted
> 
> ```
> INFO  2023-09-14 05:15:06,005 [shard 0:comp] compaction - [Compact ks.cf aa9ebb50-52bd-11ee-8637-ad387c44b32e] Compacting of 1 sstables interrupted due to: sstables::compaction_stopped_exception (Compaction for ks/cf was stopped due to: shutdown)
> INFO  2023-09-14 05:15:06,005 [shard 0:comp] compaction_manager - Compaction task 0x600001412710 for table ks.cf compaction_group=0/1 [0x60000147cd40]: Compaction for ks/cf was stopped due to: shutdown: stopping
> ```
> 
> A simple correction is about waiting for compaction to cease, before making the checks.

@raphaelsc is ccm implementation of `wait_for_compactions` would be enough ? 
or it's need to be something more then that ?

**Comment by tchaikov** [comment]
@raphaelsc hi Raphael, thank you for your great analysis. created a fix at [#3622](https://github.com/scylladb/scylla-dtest/pull/3622)

**Comment by tchaikov** [comment]
> > the dtest is built on the assumption that LCS has level L 10x bigger than L-1 as an invariant (meaning that would be the layout at any point in time), but that's not true today. That layout is eventually reached through promotions, but the dtest is not giving compaction enough time to complete its work.
> > the log shows compaction was interrupted
> > ```
> > INFO  2023-09-14 05:15:06,005 [shard 0:comp] compaction - [Compact ks.cf aa9ebb50-52bd-11ee-8637-ad387c44b32e] Compacting of 1 sstables interrupted due to: sstables::compaction_stopped_exception (Compaction for ks/cf was stopped due to: shutdown)
> > INFO  2023-09-14 05:15:06,005 [shard 0:comp] compaction_manager - Compaction task 0x600001412710 for table ks.cf compaction_group=0/1 [0x60000147cd40]: Compaction for ks/cf was stopped due to: shutdown: stopping
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > A simple correction is about waiting for compaction to cease, before making the checks.
> 
> @raphaelsc is ccm implementation of `wait_for_compactions` would be enough ? or it's need to be something more then that ?

i am afraid that it's not enough. probably you could take a look at my fix.
